app/src/functions/notify-endpoint-usage/app.ts

```diff
@@ -55,17 +57,42 @@ function getRelaysUsed(networkData: QueryAppResponse[], influxData: GetUsageData
   return applicationsData
 }
 
-exports.handler = async () => {
-  const usage = await getUsageData();
+async function getNetworkData(influxData: GetUsageDataQuery[]): Promise<QueryAppResponse[]> {
+  const sleep = (seconds: number, factor: number) => new Promise(resolve => setTimeout(resolve, (seconds ** factor) * 1000))
+
+  const networkApps: QueryAppResponse[] = []
+  const failedApps: GetUsageDataQuery[] = []
 
-  const networkApps = Promise.allSettled(
-    usage.map((app) => getApplicationNetworkData(app.applicationPublicKey))
-  );
+  for (let i = 0; i < maxRetries; i++) {
+    if (i > 0) {
+      await sleep(2, i)
+    }
+
+    const appsToQuery = i === 0 ? influxData : failedApps
+    const networkResponse = await Promise.allSettled(
+      appsToQuery.map((app) => getApplicationNetworkData(app.applicationPublicKey))
+    );
+
+    networkResponse.forEach((app, idx) => {
+      if (app.status === 'fulfilled' && app.value !== undefined) {
+        networkApps.push(app.value)
+      } else {
+        failedApps.push(appsToQuery[idx])
+      }
+    })
 
-  // TODO: Retry on error
+    if (networkApps.length === influxData.length) {
+      break
+    }
+  }
```

Comment:
This is good and obvious enough if you take into account the `for` condition above, but my grip with this is that the logic for retrying is a bit all over the places—it doesn't feel cohesive, but rather, I have to "imagine" where the code that doesn't deal with a retry goes and what's the condition for retrying.

I think that recursively applying the function with a `setTimeout` would be a tad cleaner—sounds counter-intuitive, but I have many examples where I've done this and it comes out looking incredibly clean. Let me know if we should explore that option, or if you think this is OK enough. I think we can make this cleaner.

Reply:
for sure, on my side I felt this could definitely be improved, even converted to a recursive function as the stack won't go any deep